File: Data Structures & Algorithms/binary-search/submission-2.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if (nums.size() == 1) {
            return target == nums[0] ? 0 : -1;
        }
        int l = 0, r = nums.size() - 1;

        while(l<=r){
            int mid = l + (r-l)/2;
            if (nums[mid] == target)
                return mid;
            else if (nums[mid] > target)
                r--;
            else
                l++;        
        } 

        return -1;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    threeSum(nums) {
        // brute force
        const len = nums.length
        let res = []
        let st = new Set()

        nums.sort()
        for (let i = 0; i < len; i++) {
            for (let j = i + 1; j < len; j++) {
                for (let k = j + 1; k < len; k++) {
                    if (nums[i] + nums[j] + nums[k] === 0)
                        st.add(JSON.stringify([nums[i], nums[j], nums[k]])) // set cannot detect duplication of array of array was pushed
                }
            }
        }

        st.forEach((item) => res.push(JSON.parse(item)))
        return res                
    }
}
